### PR TITLE
[NPUW] Extend HOST_GATHER with an additional case

### PR DESCRIPTION
**Pull request by smirnov-alexey**
E-176763
E-177741

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/util.cpp [hunk 1]
```diff
@@ -402,6 +432,49 @@ void ov::npuw::util::gather(const ov::SoPtr<ov::ITensor>& src,
     }
 }
 
+void ov::npuw::util::gather_nf4(const ov::SoPtr<ov::ITensor>& src,
+                                const ov::SoPtr<ov::ITensor>& idx,
+                                const ov::SoPtr<ov::ITensor>& dst) {
+    const auto src_type = src->get_element_type();
+    const auto dst_type = dst->get_element_type();
+    NPUW_ASSERT(idx->get_element_type() == ov::element::u4);
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/util.cpp:440]: that's not really NF4 but CB4, a LUT type - so I'd propose to rename the whole function to `gather_cb4`.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.hpp [hunk 2]
```diff
@@ -185,6 +192,12 @@ class HostGatherDQ : public ov::pass::MatcherPass {
     HostGatherDQ(Context::Ref ctx);
 };
 
+class HostGatherNF4 : public ov::pass::MatcherPass {
+public:
+    OPENVINO_MATCHER_PASS_RTTI("npuw::patterns::opt::HostGatherNF4");
+    HostGatherNF4(Context::Ref ctx);
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.hpp]: ex-pli-cit

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp [hunk 3]
```diff
@@ -360,9 +360,69 @@ Convert Convert::deserialize(std::istream& stream) {
...
+    }
+    seed ^= type.hash() + 0x9e3779b9;
+    for (const auto& dim : shape) {
+        seed ^= std::hash<std::size_t>()(dim) + 0x9e3779b9;
+    }
+    return seed;
+}
+
+bool Gather::operator==(const Gather& other) const {
+    // FIXME: compare tensor memory
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp]: Need to address this. E.g. save original Constant's data ptr

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp [hunk 4]
```diff
@@ -360,9 +360,69 @@ Convert Convert::deserialize(std::istream& stream) {
...
+    w.detach();
+}
+
+void Gather::serialize(std::ostream& stream) const {
+    using namespace ov::npuw::s11n;
+    write(stream, type.to_string());
+    write(stream, shape);
+    write(stream, w);
+    write(stream, t);
+}
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp]: Technically it is worth bumping a serialization version here

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp (hunk 3 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp [hunk 5]
```diff
@@ -360,9 +360,70 @@ Convert Convert::deserialize(std::istream& stream) {
...
+    using namespace ov::npuw::s11n;
+    Gather g;
+    std::string type_str;
+    read(stream, type_str);
+    g.dst_type = ov::element::Type(type_str);
+    read(stream, g.dst_shape);
+    read(stream, g.w);
+    read(stream, g.t);
+    return g;
+}
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:441]: So should you bump the serialization tag here or not?
Previously you didn't write lazy tensors of this type but now you do

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.hpp [hunk 6]
```diff
@@ -49,6 +51,11 @@ class LazyTensor {
                const LazyTensor& cs,
                const ov::element::Type& type,
                const ov::Shape& shape);  // construct from unpack
+    LazyTensor(const LazyTensor& cw,
+               const ov::Tensor& t,
+               const ov::element::Type& dst_type,
+               const ov::Shape& dst_shape,
+               const void* orig_lut_ptr);  // construct from nf4_gather
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.hpp]: why do you need this?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp [hunk 7]
```diff
@@ -2002,6 +2003,34 @@ void Partitioner::optimize(const std::string& func_name) {
...
+            const auto& tensor_to_gather = p.second;
+            auto w_idx = f._model->get_parameter_index(tensor_to_gather.w);
+
+            // Need to add a new parameter right away, since it's going to be processed by the unpack below
+            f._model->add_parameters({p.first});
+            to_remove.push_back(tensor_to_gather.w);
+            to_remove_idx.insert(w_idx);
+
+            ov::npuw::util::non_parallel_for(func_group.refs.size(), [&](std::size_t f_idx) {
+                auto& funcall = func_group.refs[f_idx].get();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:2017]: just a plain range-based for loop works here - as you don't really need `f_idx`?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 8]
```diff
@@ -109,6 +109,24 @@ Context::PPtr Context::unpack(const Context::PPtr& w, const Context::PPtr& s, ov
     return new_param;
 }
 
+Context::PPtr Context::gather_nf4(const Context::PPtr& w,
+                                  const ov::Tensor& t,
+                                  ov::element::Type type,
+                                  const void* orig_lut_ptr) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: why do you pass it everywhere

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 9]
```diff
@@ -1653,6 +1672,63 @@ HostGatherDQ::HostGatherDQ(Context::Ref ctx) {
     register_matcher(std::make_shared<opp::Matcher>(qmul, "HostGatherDQ"), std::move(callback));
 }
 
+HostGatherNF4::HostGatherNF4(Context::Ref ctx) {
+    auto pids = opp::wrap_type<ov::op::v0::Parameter>();
+    auto cvtids = opp::wrap_type<ov::op::v0::Convert>({pids});
+
+    auto qweight = opp::wrap_type<ov::op::v0::Parameter>();
+    auto qcvtw = opp::wrap_type<ov::op::v0::Convert>({qweight});
+    auto qweightt = opp::wrap_type<ov::op::v0::Constant>();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: if you use `qweightt` with extra `t` to denote a table, please rename it to `qtable` - as now it looks like a typo given there's also `qweight` few lines above (and both are used in the callback)

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp [hunk 10]
```diff
@@ -1653,6 +1672,63 @@ HostGatherDQ::HostGatherDQ(Context::Ref ctx) {
...
+            auto matched_qweight = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_qweight);
+            auto matched_qcoeff = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_qcoeff);
+            auto matched_ids = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_ids);
+            auto matched_qweightt = std::static_pointer_cast<ov::op::v0::Constant>(matched_node_qweightt);
+
+            // Need to gather the weight into f16 first
+            auto fp16weight = ctx.get().gather_nf4(matched_qweight,
+                                                   ov::npuw::util::copy_tensor_from_const(matched_qweightt),
+                                                   ov::element::f16,
+                                                   matched_qweightt->get_data_ptr());
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: so.. what's the reason to make copy right here? why don't you pass the tensor itself like it was supposed to be done?

if you just pass the tensor, you don't need to pass this pointer everywhere.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.hpp (hunk 2 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.hpp]: We don't have it for every other constructor in this file

File: src/plugins/intel_npu/src/plugin/npuw/util.cpp (hunk 1 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/util.cpp:440]: Done, renamed all gather nf4 functions and patterns to cb4

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp (hunk 4 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp (hunk 5 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:441]: Done

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.hpp (hunk 6 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.hpp]: If you are asking about the lut ptr here, I use it in hash and operator==. So it's matched properly in the bank. I could remove it since it's for vocab, but I see no harm of having it

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp (hunk 7 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/partitioning.cpp:2017]: We use `ov::npuw::util::non_parallel_for(func_group.refs.size(), [&](std::size_t f_idx) {` for all other transformations in `optimize()` - unpack, concat, permute, etc. I'd keep it consistent

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 8 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: To use in hash and operator==

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 10 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: I copy const as tensor here, so it's easier to handle `detach` and s11n. It's like 16 elements anyways. As for the pointer - I intentionally use original const ptr here, so it's shared between the models

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 9 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.hpp (hunk 6 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.hpp]: Removed, using tensor data instead

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 8 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: Removed, using tensor data instead

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp (hunk 10 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp]: Removed, using tensor data instead

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp [hunk 11]
```diff
@@ -360,9 +360,88 @@ Convert Convert::deserialize(std::istream& stream) {
...
+        seed ^= std::hash<std::size_t>()(dim) + 0x9e3779b9;
+    }
+    auto ttype = t.get_element_type();
+    NPUW_ASSERT(ttype == ov::element::f8e4m3 || ttype == ov::element::f8e5m2 || ttype == ov::element::f8e8m0);
+    std::vector<uint8_t> t_data(t.get_size());
+    std::memcpy(t_data.data(), static_cast<uint8_t*>(t.data()), t.get_size());
+    seed ^= t_data.size();
+    for (const auto& el : t_data) {
+        seed ^= std::hash<uint8_t>()(el) + 0x9e3779b9;
+    }
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:377]: you could simply take a pointer to data and iterate over the elements without an extra copy.
how often those `hash()` functions are called? could it case a performance issue in the future?

Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:377]: Worth adding that to your follow-up task

Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:377]: Created a follow-up task

Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:377]: It's called once for the model and it's just 16 elements

**Comment by dmatveev** [comment]
build_jenkins

**Comment by smirnov-alexey** [comment]
build_jenkins
